matcalc/neb.py

```diff
@@ -0,0 +1,137 @@
+
+        self.optimizer: Optimizer = getattr(optimize, optimizer) if isinstance(optimizer, str) else optimizer
+        self.traj_folder = traj_folder
+        self.interval = interval
+        self.climb = climb
+
+        self.images = []
+        for image in images:
+            if isinstance(image, Structure):
+                atoms = AseAtomsAdaptor().get_atoms(image)
+            atoms.calc = get_universal_calculator(self.calculator)
+            self.images.append(atoms)
+
+        self.neb = NEB(
+            self.images,
+            climb=self.climb,
+            allow_shared_calculator=True,
+            **kwargs,
+        )
+        self.optimizer = self.optimizer(self.neb)
+
+    @classmethod
+    def from_end_images(
+        cls,
+        start_struct: Structure,
+        end_struct: Structure,
+        calculator: str | Calculator = "M3GNet-MP-2021.2.8-DIRECT-PES",
+        nimages: int = 7,
+        interpolate_lattices: bool = False,
+        autosort_tol: float = 0.5,
+        **kwargs,
+    ):
+        """
+        Initialize a NEBCalc from end images.
+
+        Args:
+            start_struct(Structure): The starting image as a pymatgen Structure.
+            end_struct(Structure): The ending image as a pymatgen Structure.
+            calculator(str|Calculator): ASE Calculator to use. Default to M3GNet-MP-2021.2.8-DIRECT-PES.
+            nimages(int): The number of intermediate image structures to create.
+            interpolate_lattices(bool): Whether to interpolate the lattices when creating NEB
+                path with Structure.interpolate() in pymatgen. Default to False.
+            autosort_tol(float): A distance tolerance in angstrom in which to automatically
+                sort end_struct to match to the closest points in start_struct. This
+                argument is required for Structure.interpolate() in pymatgen.
+                Default to 0.5.
+            kwargs: Other arguments passed to construct NEBCalc.
+        """
+        images = start_struct.interpolate(
+            end_struct,
+            nimages=nimages + 1,
+            interpolate_lattices=interpolate_lattices,
+            pbc=False,
+            autosort_tol=autosort_tol,
+        )
+        return cls(images=images, calculator=calculator, **kwargs)
+
+    def calc(
+        self,
+        fmax: float = 0.1,
+        max_steps: int = 1000,
+    ):
+        """
+        Perform NEB calculation.
+
+        Args:
+            fmax (float): Convergence criteria for NEB calculations defined by Max forces.
+                Default to 0.1 eV/Angstrom.
+            max_steps (int): Maximum number of steps in NEB calculations. Default to 1000.
+
+        Returns:
+            NEB barrier.
+        """
+        if self.traj_folder is not None:
+            os.makedirs(self.traj_folder, exist_ok=True)
+            for i in range(len(self.images)):
+                self.optimizer.attach(
```

Comment:
We need some more tests for uncovered lines (see the codecov comments).

![Screenshot 2023-12-11 at 9 53 38 AM](https://github.com/materialsvirtuallab/matcalc/assets/30958850/37eb7308-9914-4cdf-a023-156f4a9f8691)